### PR TITLE
Fix observation normalization

### PR DESCRIPTION
**Pull request by AdamGleave**
*Changes*

1) The current implementation of AdversarialTrainer only normalizes the `obs` component, not the `next_obs` component.

    Our base reward networks don't use `next_obs` by default, so this is not as bad as it seems, but we do use it for potential shaping. This is problematic since this inconsistent normalization makes it no longer be potential shaping, and does actually change the optimal policy.

2) `VecNormalize` was applied *after* reward wrapping. So `DiscrimNet.predict_train_reward` gets called on un-normalized observations. Yet we train the discriminator on normalized observations!

    I've swapped the order, so we normalize observations first. I introduce another `VecNormalize` to normalize (and clip) rewards but not observations after the wrapped reward.

3) I added keyword arguments that let you control whether to normalize observations or rewards. There are some environments in particular where normalizing observations is unnecessary or undesirable (e.g. Atari).

4) I deleted `venv_test` since it is not used anywhere and complicates the code. I think it is plausible that we should in fact be using it somewhere, it looks like `train_adversarial` used to use it for the final rollout stage but that got swapped to `venv_train` sometime during the port.

5) I also deleted `device` in `AdversarialTrainer` which is unused. Maybe we should make it do something, but if it does nothing its inclusion is just confusing.

*Benchmark*

Running `train_adversarial` with the default config on CartPole, I got:
  - Before the change
    - GAIL: Seed 1: 379 SD 33.5. Seed 2: 500 SD 0.0. Seed 3: 381 SD 22.8.
    - AIRL: Seed 1: 8.73 SD 0.54. Seed 2: 8.84 SD 0.49. Seed 3: 9.17 SD 0.76.
  - After the change
    - GAIL: Seed 1: 438 SD 26. Seed 2: 500 SD 0.0. Seed 3: 500.0 SD 0.0.
    - AIRL: Seed 1: 9.5 SD 0.70 (but it did peak at around ~100 part way through training). Seed 2: 497 SD 12.8. Seed 3: 18.52 SD 6.5.

Of course `CartPole-v1` is a stupid evaluation environment since if the discriminator happens to be the right sign then RL will learn the right policy. (I would not be surprised if this is why it works with AIRL some of the time given how high variance this is.) Still, the changes have certainly not made training performance any worse, and it seems like it may be a bit more stable.

I tried to get it working with `seals/CartPole-v0` which is fixed horizon but couldn't. Might just need more timesteps or other hyperparameters tweaked. Since even RL training on the ground-truth is a lot slower in that environment.

### CODE REVIEW

**Comment by AdamGleave** [comment]
OK, take those benchmarks with a gigaton of salt. I found that by flipping the sign in `DiscrimNet._eval_reward` I can make AIRL reliably get >300 and GAIL <20. Who knew the sign would be such an important hyperparameter!

I think we really need to benchmark in some fixed-horizon tasks to get anything meaningful. I wonder if we should even refuse to run on variable-horizon environments unless the user sets the `shoot_self_in_foot` flag to true?

**Comment by AdamGleave** [comment]
Results from ```parallel --header : python -m imitation.scripts.train_adversarial with env_name=seals/CartPole-v0 rollout_path=data/expert_models/cartpole_0/rollouts/final.pkl total_timesteps=1e6 {algo} seed={seed} log_dir={algo}/{seed}/ ::: algo gail airl ::: seed 0 1 2 3 4``` on current master (https://github.com/HumanCompatibleAI/imitation/commit/23e7c150074f522715ef839d6609175076cb9d0f) and this PR (https://github.com/HumanCompatibleAI/imitation/pull/256/commits/343eb25c551909c92af50a6c6d0162d4b31e1cb1)

Reporting mean ground-truth return means and SDs (`monitor_return_mean` and `monitor_return_sd`) over 56 imitation trajectories. (So S.E. is about 1/7th the SD.)

master GAIL (before this change)

| Seed | Mean | SD    |
| ------- | -------- | ------ |
| 0       | 500.0  | 0.0   |
| 1       | 11.93  | 0.92 |
| 2       | 500.0  | 0.0   |
| 3       | 22.4    | 9.58 |
| 4       | 500.0  | 0.0   |

image


This PR GAIL

| Seed | Mean | SD    |
| ------- | -------- | ------ |
| 0       | 500     | 0.0   |
| 1       | 480.8  | 19.76 |
| 2       | 500.0  | 0.0   |
| 3       | 500.0  | 0.0   |
| 4       | 500.0  | 0.0   |

image

master AIRL

| Seed | Mean | SD    |
| ------- | -------- | ------ |
| 0       | 8.52    | 0.78 |
| 1       | 456.9  | 59.9 |
| 2       | 500.0  | 0.0   |
| 3       | 8.5      | 0.96 |
| 4       | 385.1  | 49.9 |

image

This PR AIRL

| Seed | Mean | SD    |
| ------- | -------- | ------ |
| 0       | 500.0  | 0.0   |
| 1       | 242.7  | 10.12 |
| 2       | 306.8  | 8.37   |
| 3       | 500.0  | 0.0   |
| 4       | 500.0  | 0.0   |

image

Full logs attached.

Changes seem to have significantly stabilized training, making the minimum mean return across seeds much higher. However, the maximum return achieved is unchanged at least in this simple environment.

Training is slow: basically flatlines for the first 100,000 timesteps. Reward oscillates during training. For GAIL the size of the oscillations seem to decrease over time. It looks like it actually oscillates more over time after this PR than before, but there's less variance between seeds. AIRL training curves look like more extreme version of GAIL. Most seeds don't work on master, all train somewhat with this PR, but one flatlines for 600,000 timesteps! Lots of oscillation, seems to be slow to decay and may have plateaued towards the end.

Overall seems like a significant improvement but I feel we should be able to do better, especially on a simple task like CartPole.

**Comment by AdamGleave** [comment]
I may try and make some further improvements in other PRs but I think this is now ready for review @qxcv @shwang

**Comment by shwang** [comment]
Great, I'm taking a look today

**Review by shwang — APPROVED** [review]
Thanks for these catches and the comparison runs -- good to see an improvement, sad to see that there were still lingering VecNormalization problems!

I've confirmed that the diff itself makes sense, that it matches the PR description, and verified the problems mentioned in the PR. 

I didn't check whether variable name changes match with things outside the diff, etc. because I'm swamped right now with grad apps stuff right now.

If it is helpful, I can give this PR more thorough look later.

**Comment by shwang** [comment]
We discussed this a while ago, but to refresh, it seems like all this normalization logic would be automatically solved for us (except reward normalization), and serialization would be greatly simplified if there were an SB3 policy that included VecNorm-style normalization?

The batchnorm (mean=0, std=1) solution you [mentioned](https://github.com/HumanCompatibleAI/imitation/issues/229#issuecomment-742539096) seems like a promising approach.

File: src/imitation/algorithms/adversarial.py [hunk 1]
```diff
@@ -23,18 +23,17 @@ class AdversarialTrainer:
...
+
+    These statistics must be saved along with the model."""
+
     venv_train: vec_env.VecEnv
     """Like `self.venv`, but wrapped with train reward unless in debug mode.
 
     If `debug_use_ground_truth=True` was passed into the initializer then
     `self.venv_train` is the same as `self.venv`."""
 
-    venv_test: vec_env.VecEnv
```
Comment (shwang) [inline src/imitation/algorithms/adversarial.py:32]: This should be fine to remove.

**Comment by AdamGleave** [comment]
Thanks for the review, especially in these busy times!

> We discussed this a while ago, but to refresh, it seems like all this normalization logic would be automatically solved for us (except reward normalization), and serialization would be greatly simplified if there were an SB3 policy that included VecNorm-style normalization?

We have three places where normalization happens:
  1. Normalization of observations for the generator (policy)
  2. Normalization of observations for the discriminator (reward, for AIRL);
  3. Normalization of reward (from discriminator).

Making a policy wrapper to normalize observations would only handle the first case. We could make a separate discriminator wrapper that normalizes observations. I think this would be meaningfully different, though. Right now observation normalization statistics are only updated on-policy from fresh samples from the environment. A discriminator wrapper would presumably instead see samples we use during training, which would be 50% expert, and could be repeated (if sampled from a replay buffer).

My guess (~80%) is both approaches would achieve similar performance, with ~10% chance that one performs better than the other (but I don't know which!)

I discussed policy normalization wrappers with SB3 maintainers a while back and it seems unlikely to make it in. But of course we can implement it in `imitation` if we think it'd be simpler for our particular use case. I think I weakly favor this but it doesn't seem a priority (compared to e.g. improving algorithm performance or better docs)

> > The batchnorm (mean=0, std=1) solution you [mentioned](https://github.com/HumanCompatibleAI/imitation/issues/229#issuecomment-742539096) seems like a promising approach.

Thanks, hope to look into it soon!

**Review by qxcv — APPROVED** [review]
Also looks good to me, although I'm also sceptical that normalising observations with the running mean is more helpful than simpler alternatives (such as rescaling using the `Box` bounds, and/or using batch/layer norm).

**Comment by AdamGleave** [comment]
> Also looks good to me, although I'm also sceptical that normalising observations with the running mean is more helpful than simpler alternatives (such as rescaling using the `Box` bounds, and/or using batch/layer norm).

Thanks for the review!

FYI I think a decent number of environments don't have Box bounds, e.g.:

```python
import gym
env = gym.make('HalfCheetah-v2')
print(env.observation_space.low)
```

gives:

```python
array([-inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf,
       -inf, -inf, -inf, -inf, -inf, -inf])
```

Admittedly `VecNormalize` is a heavy weight solution to that problem. (You could also just renormalize based on static statistics, e.g. from the expert demos or a rollout of a random policy.)
